src/history/base.rs

```diff
@@ -14,42 +20,285 @@ pub enum HistoryNavigationQuery {
+        }
+    }
+    pub fn last_with_search(filter: SearchFilter) -> SearchQuery {
+        SearchQuery {
+            direction: SearchDirection::Backward,
+            start_time: None,
+            end_time: None,
+            start_id: None,
+            end_id: None,
+            limit: Some(1),
+            filter,
+        }
+    }
+    pub fn last_with_prefix(prefix: String) -> SearchQuery {
+        SearchQuery::last_with_search(SearchFilter::from_text_search(CommandLineSearch::Prefix(
+            prefix,
+        )))
+    }
+    pub fn everything(direction: SearchDirection) -> SearchQuery {
+        SearchQuery {
+            direction,
+            start_time: None,
+            end_time: None,
+            start_id: None,
+            end_id: None,
+            limit: None,
+            filter: SearchFilter::anything(),
+        }
+    }
+}
+
+/// Represents a history file or database
+/// Data could be stored e.g. in a plain text file, in a JSONL file, in a SQLite database
 pub trait History: Send {
-    /// Append entry to the history, if capacity management is part of the implementation may perform that as well
-    fn append(&mut self, entry: &str);
+    /// save a history item to the database
+    /// if given id is None, a new id is created and set in the return value
+    /// if given id is Some, the existing entry is updated
+    fn save(&mut self, h: HistoryItem) -> Result<HistoryItem>;
+    /// load a history item by its id
+    fn load(&self, id: HistoryItemId) -> Result<HistoryItem>;
+    /// gets the newest item id in this history
+    fn newest(&self) -> Result<Option<HistoryItemId>> {
+        let res = self.search(SearchQuery::last_with_search(SearchFilter::anything()))?;
+        Ok(res.get(0).and_then(|e| e.id))
+    }
 
-    /// Chronologic interaction over all entries present in the history
-    fn iter_chronologic(&self) -> Iter<'_, String>;
+    /// creates a new unique session id
+    fn new_session_id(&mut self) -> Result<HistorySessionId>;
 
-    /// This moves the cursor backwards respecting the navigation query that is set
-    /// - Results in a no-op if the cursor is at the initial point
-    fn back(&mut self);
+    /// count the results of a query
+    fn count(&self, query: SearchQuery) -> Result<i64>;
+    /// return the total number of history items
+    fn count_all(&self) -> Result<i64> {
+        self.count(SearchQuery::everything(SearchDirection::Forward))
+    }
+    /// return the results of a query
+    fn search(&self, query: SearchQuery) -> Result<Vec<HistoryItem>>;
 
-    /// This moves the cursor forwards respecting the navigation-query that is set
-    /// - Results in a no-op if the cursor is at the latest point
-    fn forward(&mut self);
+    /// update an item atomically
+    fn update(
+        &mut self,
+        id: HistoryItemId,
+        updater: &dyn Fn(HistoryItem) -> HistoryItem,
+    ) -> Result<()>;
```

Comment:
Thinking about it: is this a capability of the base history trait? (would it make sense to lift `Reedline::update_last_command_context` to the concrete history type or a subtype trait `ExtendableHistory`? [gnarf:  can we export it from `Reedline` while it has to have a concrete `dyn History` trait bound to exist in the struct (in a type generic way it would be possible but messy)])

Reply:
That's actually pretty much how I had it in #376. The main issue there was that the outside application had to retain a handle to the SqliteHistory so I had to wrap it in a Mutex. This is better now since `update` is part of the History trait itself, so the .history() (or `.history_mut()` function could work. If you want a more concrete trait then some downcasting needs to happen..

Another issue is that the update_last_command_context requires state (the `last_run_command_id` currently stored in Reedline), which the History trait doesn't know (and can't reconstruct because multiple shells could be open).

Reply:
Sorry, I kind of misread your comment. Multiple parts of the history trait (e.g. parts of the search params) only work on some History types, but I think it would be fairly hard to separate those out (instead of failing at runtime as it does now).

Reply:
Yeah unless we completely switch over to compiled generics, we will have to do runtime dispatch in some form (if we also want to be able to hotswap some components). This was just spitballing if we could get it a bit neater but for now going for effectively `NotImplementedError`s should be fine